### PR TITLE
refactor

### PR DESCRIPTION
**Pull request by defbin**

### CODE REVIEW

File: pkg/orchestrator/containers.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
+package orchestrator
```
Comment (egegunes) [inline pkg/orchestrator/containers.go:1]: if we're going to put together all functions in a single file in `mysql` package, let's be consistent and do the same for `orchestrator`

File: pkg/reconcile/secrets.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+const (
+	passwordMaxLen = 20
+	passwordMinLen = 16
+	passSymbols    = "ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+		"abcdefghijklmnopqrstuvwxyz" +
+		"0123456789"
+)
+
+// generatePass generates a random password
+func generatePass() ([]byte, error) {
```
Comment (egegunes) [inline pkg/reconcile/secrets.go:21]: maybe move this code to the new `util` package?

File: pkg/reconcile/reconcile.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
+package reconcile
```
Comment (egegunes) [inline pkg/reconcile/reconcile.go:1]: I don't think this new package name is any better than the old one. We have 3 different controllers and all of them do the reconciliation.

File: pkg/mysql/mysql.go [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+func ServiceName(cr *v2.PerconaServerForMySQL) string {
+	return Name(cr)
+}
+
+func PrimaryServiceName(cr *v2.PerconaServerForMySQL) string {
+	return Name(cr) + "-primary"
+}
+
+// func IsMySQL(obj client.Object) bool {
```
Comment (egegunes) [inline pkg/mysql/mysql.go:43]: let's remove the commented code, they're not used

**Review by egegunes — CHANGES_REQUESTED** [review]

File: controllers/psmdb/reconcile.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
+package psmdb
```
Comment (egegunes) [inline controllers/psmdb/reconcile.go:1]: psmdb?

File: controllers/psmdb/reconcile.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+	if err := ensureObject(ctx, cr, secret, c.client.Scheme(), c.client.Create); err != nil {
+		return errors.Wrapf(err, "create secret %s", k8s.SecretsName(cr))
+	}
+
+	return nil
+}
+
+var secretUsers = [...]string{
+	apiv2.USERS_SECRET_KEY_ROOT,
```
Comment (egegunes) [inline controllers/psmdb/reconcile.go:98]: Can we shorten these variable names? For example `apiv2.UsersRoot`

File: controllers/psmdb/reconcile.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+}
+
+func ensureObject(
+	ctx context.Context,
+	cr *apiv2.PerconaServerForMySQL,
+	o client.Object,
+	s *runtime.Scheme,
+	create createObjectFn,
+) error {
+	if err := k8s.SetControllerReference(cr, o, s); err != nil {
```
Comment (egegunes) [inline controllers/psmdb/reconcile.go:203]: Let's use https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#SetControllerReference

File: controllers/psmdb/replication.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+	operatorPass, err := k8s.UserPassword(ctx, rdr, cr, v2.USERS_SECRET_KEY_OPERATOR)
+	if err != nil {
+		return errors.Wrap(err, "get operator password")
+	}
+
+	db, err := replicator.NewReplicator(v2.USERS_SECRET_KEY_OPERATOR,
+		operatorPass,
+		primary.Hostname(),
+		int32(33062))
```
Comment (egegunes) [inline controllers/psmdb/replication.go:69]: `mysql.DefaultAdminPort`

File: pkg/k8s/cr.go [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package k8s
+
+import psmdbv2 "github.com/percona/percona-server-mysql-operator/api/v2"
```
Comment (egegunes) [inline pkg/k8s/cr.go:3]: psmdbv2?

File: pkg/k8s/cr.go [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+	return cr.Namespace
+}
+
+func SecretsName(cr *psmdbv2.PerconaServerForMySQL) string {
+	return cr.Spec.SecretsName
+}
+
+func SSLSecretName(cr *psmdbv2.PerconaServerForMySQL) string {
+	return cr.Spec.SSLSecretName
+}
```
Comment (egegunes) [inline pkg/k8s/cr.go:15]: I don't understand what is the point of having these functions?

File: pkg/util/map.go2 [hunk 11]
```diff
@@ -0,0 +1,42 @@
+package util
```
Comment (egegunes) [inline pkg/util/map.go2:1]: It seems like a leftover

**Review by egegunes — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/80ae820c84bcc188dd10d8b840940d71c116c757
image: `perconalab/percona-server-mysql-operator:PR-18-80ae820`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/218db1de8ffa911a591dbfb6305299a722c24200
image: `perconalab/percona-server-mysql-operator:PR-18-218db1d`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/48aa636acbd996ca3abac6f21ff5b381363bcbea
image: `perconalab/percona-server-mysql-operator:PR-18-48aa636`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/a31da5f551f4e9880473007415be7b83ea165a7f
image: `perconalab/percona-server-mysql-operator:PR-18-a31da5f`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/d2a319f1b03430c8aa6e635fc5c4c1c1fdf325ef
image: `perconalab/percona-server-mysql-operator:PR-18-d2a319f`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/e412e0ac30d7491f1cd38ec895c409c5184ca485
image: `perconalab/percona-server-mysql-operator:PR-18-e412e0a`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/85f688b72560d6c72a9b4dc800a63fb49df9578a
image: `perconalab/percona-server-mysql-operator:PR-18-85f688b`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/6b6d697c5691c6fb31b7ad6f25d80c98327d9102
image: `perconalab/percona-server-mysql-operator:PR-18-6b6d697`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/843a21222cb2831849ad4bb5ceefe57f247c3759
image: `perconalab/percona-server-mysql-operator:PR-18-843a212`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/8d789feec91073e2c25244ae4378f7635c70d5ec
image: `perconalab/percona-server-mysql-operator:PR-18-8d789fe`

File: controllers/perconaserverformysql_controller.go [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+}
+
+func (r *PerconaServerForMySQLReconciler) loadCR(ctx context.Context, nn types.NamespacedName) (*apiv2.PerconaServerForMySQL, error) {
+	o, err := k8s.GetObjectWithDefaults(ctx, r.Client, nn, &apiv2.PerconaServerForMySQL{})
+	if err != nil {
+		return nil, err
+	}
+
+	return o.(*apiv2.PerconaServerForMySQL), nil
+}
```
Comment (delgod) [inline controllers/perconaserverformysql_controller.go:120]: function has no sense, I prefer to run `GetObjectWithDefaults` directly or put `GetObjectWithDefaults` logic inside this function

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/b3122f00cf4fa683dd365c9100ae8f09e92b972a
image: `perconalab/percona-server-mysql-operator:PR-18-b3122f0`

File: api/v2/perconaserverformysql_types.go [hunk 13]
```diff
@@ -33,14 +32,14 @@ import (
...
-	SSLSecretName         string    `json:"sslSecretName,omitempty"`
-	SSLInternalSecretName string    `json:"sslInternalSecretName,omitempty"`
-	MySQL                 MySQLSpec `json:"mysql,omitempty"`
-	Orchestrator          PodSpec   `json:"orchestrator,omitempty"`
-	PMM                   *PMMSpec  `json:"pmm,omitempty"`
+	CRVersion             string           `json:"crVersion,omitempty"`
+	Pause                 bool             `json:"pause,omitempty"`
+	SecretsName           string           `json:"secretsName,omitempty"`
+	SSLSecretName         string           `json:"sslSecretName,omitempty"`
+	SSLInternalSecretName string           `json:"sslInternalSecretName,omitempty"`
```
Comment (egegunes) [inline api/v2/perconaserverformysql_types.go:39]: We have the same fields in `PodSpec` struct. Should we keep them both? @delgod @hors

Currently we're using only these fields.

File: controllers/perconaserverformysql_controller.go [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, errors.Wrap(err, "get CR")
+	}
+
+	if err := r.doReconcile(ctx, l, cr); err != nil {
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, err
```
Comment (egegunes) [inline controllers/perconaserverformysql_controller.go:86]: let's wrap the error

File: controllers/perconaserverformysql_controller.go [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+	return nil
+}
+
+func (r *PerconaServerForMySQLReconciler) getCRWithDefaults(
+	ctx context.Context,
+	nn types.NamespacedName,
+) (*apiv2.PerconaServerForMySQL, error) {
+	cr := &apiv2.PerconaServerForMySQL{}
+	if err := r.Client.Get(ctx, nn, cr); err != nil {
+		return nil, err
```
Comment (egegunes) [inline controllers/perconaserverformysql_controller.go:122]: please wrap errors

File: controllers/perconaserverformysql_controller.go [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+		if pods[i].Name == primaryAlias {
+			patch := client.StrategicMergeFrom(pod)
+			k8s.AddLabel(pod, apiv2.MySQLPrimaryLabel, "true")
+			if err := cl.Patch(ctx, pod, patch); err != nil {
+				return errors.Wrap(err, "add label to new primary pod")
+			}
+
+			break
+		}
+	}
```
Comment (egegunes) [inline controllers/perconaserverformysql_controller.go:290]: why don't we just `Get` the pod with name `primaryAlias`?

File: pkg/k8s/init_image.go [hunk 17]
```diff
@@ -9,49 +9,46 @@ import (
...
+		return nil, err
 	}
 
-	return operatorPod, nil
+	return pod, nil
 }
 
-func InitImage(cl client.Client, cr *v2.PerconaServerForMySQL) (string, error) {
-	pod, err := operatorPod(cl)
+func DefaultAPINamespace() (string, error) {
```
Comment (egegunes) [inline pkg/k8s/init_image.go:47]: what does `API` mean in this function name?

File: controllers/perconaserverformysql_controller.go (hunk 16 above)
Comment (defbin) [reply controllers/perconaserverformysql_controller.go:290]: we still need the list of pods here to remove label from old primary

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mysql-operator/pull/18/commits/321a08078181596a42a4c8ebe2e2d2d91785ad34
image: `perconalab/percona-server-mysql-operator:PR-18-321a080`

File: api/v2/perconaserverformysql_types.go (hunk 13 above)
Comment (delgod) [reply api/v2/perconaserverformysql_types.go:39]: 1. I am not sure why we need fields in `PodSpec`. there are should be only one place for specifying certificates.
2. we should use internal certificates in `clone_ssl_*` options

**Review by delgod — APPROVED** [review]

**Review by egegunes — APPROVED** [review]

**Pull request merged**